bazel/linux/test.bzl

```diff
@@ -4,6 +4,83 @@ load("//bazel/utils:macro.bzl", "mconfig", "mcreate_rule")
 load("//bazel/utils:exec_test.bzl", "exec_test")
 load("//bazel/linux:bundles.bzl", "kunit_bundle")
 
+def _test_runner(ctx):
+    script_begin = """#!/bin/bash
+set -eu -o pipefail
+GREEN="\033[32m"
+RED="\033[31m"
+RESET="\033[0m"
+failed_outputs=()
+tests=0
+failed=0
+run_test() {
+    local title="$1"
+    local script="$2"
+    local success=0
+    local output
+    local status
+    local start_time_ns
+    local stop_time_ns
+    local duration_ns
+
+    printf "%-73s" "${title}"
+    start_time_ns=$(($(date +%s%N)))
+
+    output=$("${script}" 2>&1) && success=1
+
+    tests=$[tests + 1]
+    if [ $success -eq 1 ]; then
+        status="${GREEN}PASSED${RESET}"
+    else
+        failed=$[failed + 1]
+        status="${RED}FAILED${RESET}"
+        failed_outputs+="==================== Test output for ${title}:\\n${output}\\n"
+    fi
+    stop_time_ns=$(($(date +%s%N)))
+    duration_ns=$(($stop_time_ns - $start_time_ns))
+    printf "%s in %.1fs\\n" "$status" "$duration_ns""e-9"
+}
+"""
+
+    script_end = """
+for o in "${failed_outputs[@]}"; do
+    echo -ne "$o"
+done
+if [ $failed -ne 0 ]; then
+    echo "================================================================================"
+fi
+succeeded=$[tests - failed]
+echo "Executed ${tests} test(s): ${succeeded} test(s) pass(es) and ${failed} fail(s)."
+exit $failed
+"""
+
+    tests = []
+    runfiles = ctx.runfiles()
+    for test in ctx.attr.tests:
+        binary = test.files_to_run.executable
```

Comment:
test here could be any target / label. It may not be executable. If it's not, user will get a stack trace very hard to read.

I'd try to be more defensive here, do something like:
```
if not test.files_to_run.executable:
    fail(location(ctx) + ... "you specified a target that is not executable " + package(test))
```
for example.

Reply:
if we do end up breaking the vm_bundle features here, I'd add a test to see if there is a bundle provider, if there is (and it has a pre/setup and check step), throw an error?

Reply:
Added the error message.